### PR TITLE
Allow the use of bearer_token or bearer_token_file for MarathonSD authorization

### PR DESCRIPTION
**Pull request by m-kraus**
This PR enables service discovery for Marathon with token based authorization.
The token specified in bearer_token or bearer_token_file is sent as request header (Authorization) to Marathon / DC/OS

@fabxc am i pinging the right one with this PR?

Best regards, Michael

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -939,6 +942,12 @@ func (c *MarathonSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) erro
 	if len(c.Servers) == 0 {
 		return fmt.Errorf("Marathon SD config must contain at least one Marathon server")
 	}
+	if len(c.BearerToken) > 0 && len(c.BearerTokenFile) > 0 {
+		return fmt.Errorf("at most one of bearer_token & bearer_token_file must be configured")
+	}
+	if len(c.BearerToken) == 0 && len(c.BearerTokenFile) == 0 {
+		return fmt.Errorf("at most one of bearer_token & bearer_token_file must be configured")
+	}
```
Comment (fabxc) [inline config/config.go]: This check forces you to configure at least one of them, no?

File: config/config.go [hunk 2]
```diff
@@ -920,6 +921,8 @@ type MarathonSDConfig struct {
 	Timeout         model.Duration `yaml:"timeout,omitempty"`
 	RefreshInterval model.Duration `yaml:"refresh_interval,omitempty"`
 	TLSConfig       TLSConfig      `yaml:"tls_config,omitempty"`
+	BearerToken     string         `yaml:"bearer_token,omitempty"`
+	BearerTokenFile string         `yaml:"bearer_token_file,omitempty"`
```
Comment (fabxc) [inline config/config.go]: These need to be considered in `resolveFilepaths()`.

File: discovery/marathon/marathon.go [hunk 3]
```diff
@@ -201,11 +213,20 @@ type AppList struct {
 }
 
 // AppListClient defines a function that can be used to get an application list from marathon.
-type AppListClient func(client *http.Client, url string) (*AppList, error)
+type AppListClient func(client *http.Client, url, token string) (*AppList, error)
```
Comment (fabxc) [inline discovery/marathon/marathon.go]: This broke the tests.

File: discovery/marathon/marathon.go [hunk 4]
```diff
@@ -201,11 +213,20 @@ type AppList struct {
...
-	resp, err := client.Get(url)
+func fetchApps(client *http.Client, url, token string) (*AppList, error) {
+	request, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if token != "" {
+		request.Header.Set("Authorization", "token="+token)
+	}
```
Comment (fabxc) [inline discovery/marathon/marathon.go]: The token should not be set as a distinct header on every requests but be part of the client on initialization.

We have helpers available for that:
https://github.com/prometheus/prometheus/blob/master/util/httputil/client.go

File: config/config.go (hunk 1 above)
Comment (m-kraus) [inline config/config.go]: You're right, it's better to omit that...

**Comment by m-kraus** [comment]
Thanks for the review, i will implement your suggestions as soon as possible.

File: config/config.go (hunk 2 above)
Comment (brian-brazil) [inline config/config.go]: Even better, use a HTTPClientConfig to avoid duplicating this code.

File: discovery/marathon/marathon.go (hunk 4 above)
Comment (m-kraus) [inline discovery/marathon/marathon.go]: I remember, why i did not use that:

httputil/client.go sets the teh Authorization header as follows:
`req.Header.Set("Authorization", "Bearer "+rt.bearerToken)`

But this seems not to work for DC/OS [https://dcos.io/docs/1.8/administration/id-and-access-mgt/managing-authentication/](url):
> Tokens that are sent to DC/OS in an HTTP Authorization header must be in this format: token=<token>. In future versions Bearer <token> will also be supported.

At least for the DC/OS environment running 1.8 where I can test against, Bearer does not work.

File: config/config.go (hunk 2 above)
Comment (m-kraus) [inline config/config.go]: It's already there in `resolveFilepaths()`:
`   mcfg.BearerTokenFile = join(mcfg.BearerTokenFile)`

It's done the same way as for KubernetesSD:
`    kcfg.BearerTokenFile = join(kcfg.BearerTokenFile)`

I will look into HTTPClientConfig as well.

File: discovery/marathon/marathon.go (hunk 4 above)
Comment (fabxc) [inline discovery/marathon/marathon.go]: I see, makes sense then. Would be best to add an appropriate comment so we know why in the future.

**Comment by m-kraus** [comment]
@fabxc Tests are now fixed and your requested changes should now also be covered.

Best regards, Michael

**Comment by fabxc** [comment]
Looks good. Thanks!.
Can you document the addition in https://github.com/prometheus/docs/blob/master/content/docs/operating/configuration.md?

**Comment by m-kraus** [comment]
@fabxc Documented with https://github.com/prometheus/docs/pull/676

**Comment by fabxc** [comment]
Thanks.

**Pull request merged**